### PR TITLE
Compiler optimizations

### PR DESCRIPTION
**Pull request by breznak**
- compiler performance optimizations, Ofast, cca 20% in SP performance
- cmake: separate compile unit nupic/algorithms, which gets optimization flags
- Cells4/TP: removal of unused code
- cmake: removed NTA_ARCH_*, NTA_INTERNAL
- SP: compute's param SDR input marked const

### CODE REVIEW

File: CommonCompilerConfig.cmake [hunk 1]
```diff
@@ -170,20 +158,13 @@ if(MSVC)
...
-	endif()
 	set(INTERNAL_LINKER_FLAGS "${machine} /NOLOGO /SAFESEH:NO /NODEFAULTLIB:LIBCMT /ignore:4099 /LTCG")
 
 	set(COMMON_COMPILER_DEFINITIONS 	
 		_CONSOLE
 		_MBCS
 		NTA_OS_WINDOWS
 		NTA_COMPILER_MSVC
-		NTA_INTERNAL=1
-		NTA_ARCH_${BITNESS}
```
Comment (breznak) [inline CommonCompilerConfig.cmake]: was able to remove NTA_INTERNAL, NTA_ARCH_bitness

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -253,40 +254,57 @@ source_group("types" FILES ${types_files})
...
-    ${regions_files} 
     ${types_files} 
     ${utils_files} 
 )
 
 
-target_compile_options( ${src_lib_static_nupiccore_solo} PUBLIC ${INTERNAL_CXX_FLAGS})
-target_compile_definitions(${src_lib_static_nupiccore_solo} PRIVATE ${COMMON_COMPILER_DEFINITIONS})
-target_include_directories(${src_lib_static_nupiccore_solo} PRIVATE 
+target_compile_options( ${src_lib_static_nupiccore_alg} PUBLIC ${INTERNAL_CXX_FLAGS} ${CXX_FLAGS_OPTIMIZED})
```
Comment (breznak) [inline src/CMakeLists.txt]: separate build units src/nupic/algorithms (gets optimization flags, Ofast) 
and rest (netoworkAPI)

File: src/nupic/algorithms/SpatialPooler.cpp [hunk 3]
```diff
@@ -503,7 +503,7 @@ void SpatialPooler::compute(const UInt inputArray[], bool learn, UInt activeArra
 }
 
 
-void SpatialPooler::compute(SDR &input, bool learn, SDR &active) {
+void SpatialPooler::compute(const SDR &input, bool learn, SDR &active) {
```
Comment (breznak) [inline src/nupic/algorithms/SpatialPooler.cpp]: thanks to now possible const SDR, I can mark compute(const SDR input,...) to designate it's only input data.

File: src/nupic/algorithms/Connections.hpp [hunk 4]
```diff
@@ -425,7 +425,8 @@ class Connections : public Serializable
    * @param increment  Change in permanence for synapses with active presynapses.
    * @param decrement  Change in permanence for synapses with inactive presynapses.
    */
-  void adaptSegment(const Segment segment, SDR &inputs,
+  void adaptSegment(const Segment segment, 
+		    const SDR &inputs,
```
Comment (breznak) [inline src/nupic/algorithms/Connections.hpp]: @ctrl-z-9000-times do you use adapt segment in CP? Otherwise it's only used in SP, so I'd like to move it there.

Comment (ctrl-z-9000-times) [reply src/nupic/algorithms/Connections.hpp]: I think this belongs here.  For code organization, I want all of the generic tools for synapses & segments in the Connections class.  
* The CP will probably use this method, after it's written.
* The TM does not use this method but with some simple modifications could use this method.

File: src/CMakeLists.txt (hunk 2 above)
Comment (ctrl-z-9000-times) [reply src/CMakeLists.txt]: Is there a reason that the optimization flags are not / should not be applied to the "rest"/networkAPI?

File: src/nupic/algorithms/Connections.hpp (hunk 4 above)
Comment (breznak) [reply src/nupic/algorithms/Connections.hpp]: sounds fair. 

> The TM does not use this method but with some simple modifications could use this method. 

might be nice if TM used Segments this way as well. For improved biological detail as well as shared codepaths.

File: src/CMakeLists.txt (hunk 2 above)
Comment (breznak) [reply src/CMakeLists.txt]: > Is there a reason that the optimization flags are not / should not be applied to the "rest"/networkAPI?

the algorithms/ folder is the only workhorse here. 

> "rest"/networkAPI?

they could, but
- neglible/negative impact
- longer compile time
- i wanted to separate hard-core optimization vs rather security checks focused rest

File: CommonCompilerConfig.cmake [hunk 5]
```diff
@@ -170,20 +158,13 @@ if(MSVC)
 							$<$<CONFIG:Release>:/O2 /Oi /Gy  /GL /MD> 
 							$<$<CONFIG:Debug>:/Ob0 /Od /Zi /sdl /RTC1 /MDd>)
 	#linker flags
-	if("${BITNESS}" STREQUAL "32")
-		set(machine "/MACHINE:X86")
-	else()
-		set(machine "/MACHINE:X${BITNESS}")
```
Comment (dkeeney) [inline CommonCompilerConfig.cmake]: I think we still need this.  64bit is not always the default on Windows.  I agree that other places we would break if we compiled for 32bit but perhaps we do want to support 32bit at some point.

File: CommonCompilerConfig.cmake (hunk 1 above)
Comment (dkeeney) [reply CommonCompilerConfig.cmake]: 👍

File: CommonCompilerConfig.cmake [hunk 6]
```diff
@@ -292,15 +271,6 @@ else()
 	set(cxx_flags_unoptimized ${cxx_flags_unoptimized} ${stdlib_common} -fdiagnostics-show-option)
 	set (internal_compiler_warning_flags ${internal_compiler_warning_flags} -Werror -Wextra -Wreturn-type -Wunused -Wno-unused-variable -Wno-unused-parameter -Wno-missing-field-initializers)
 
-	CHECK_CXX_COMPILER_FLAG(-m${BITNESS} compiler_supports_machine_option)
```
Comment (dkeeney) [inline CommonCompilerConfig.cmake]: I think we want to keep bitness and arm.  We eventually want to build for Rasberry Pi and such.

File: src/CMakeLists.txt (hunk 2 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: I disagree.  networkAPI needs optimization as much as algorithms.  networkAPI does not use as much resources as algorithms but every little bit helps.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -253,40 +254,57 @@ source_group("types" FILES ${types_files})
...
+target_compile_definitions(${src_lib_static_nupiccore_alg} PRIVATE ${COMMON_COMPILER_DEFINITIONS})
```
Comment (dkeeney) [inline src/CMakeLists.txt]: I am in the process of changing all of this for shared library.

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -253,40 +254,57 @@ source_group("types" FILES ${types_files})
...
+endif()
+
 #--------------------------------------------------------
 #
 # Create the nupic_core as a "combined" static library by merging nupic_core_solo
 # with our external static libraries
 #
 set(src_combined_nupiccore_source_archives
-    ${src_lib_static_nupiccore_solo}
+    ${src_lib_static_nupiccore_alg}
```
Comment (dkeeney) [inline src/CMakeLists.txt]: I see no reason to separate them.

File: src/CMakeLists.txt (hunk 2 above)
Comment (breznak) [reply src/CMakeLists.txt]: > I disagree. networkAPI needs optimization as much as algorithms. networkAPI does not use as much resources as algorithms but every little bit helps. 

Ok, what's true...we can't decide now, there's no functioning NetworkAPI benchmark. We should test with the NetworkAPI version of Hotgym benchmark, when all parts are available. 

I suspect compiler optimizations won't matter there much, as NAPI constructs the network, moves data around (ideally pointers) and instruments execution of the algs (optimized here). 

See what `Ofast` enables, it may help performance, but also may harm it, those features are highly task specific. 

TL;DR: we should revisit with Networkapi hotgym benchmark and choose the best options for each. (Ofast does not necessarily mean best)

File: CommonCompilerConfig.cmake (hunk 5 above)
Comment (breznak) [reply CommonCompilerConfig.cmake]: Do I need to revert and use NTA_ARCH_32,... or this block and $BITNESS should be enough for Win? 

>  but perhaps we do want to support 32bit at some point. 

reallistically I don't think so, major distros are dropping 32bit support (Ubuntu, fedora)

File: CommonCompilerConfig.cmake (hunk 6 above)
Comment (breznak) [reply CommonCompilerConfig.cmake]: wouldn't the `-marm` be auto recognized by cmake?

File: src/CMakeLists.txt (hunk 7 above)
Comment (breznak) [reply src/CMakeLists.txt]: still that split should be available (?)

File: src/CMakeLists.txt (hunk 8 above)
Comment (breznak) [reply src/CMakeLists.txt]: more modularity to give you option to use the optimal flags (speed, code stability, checks,..) for each of them.

File: CommonCompilerConfig.cmake (hunk 5 above)
Comment (dkeeney) [reply CommonCompilerConfig.cmake]: I see postings in the forums talking about trying to get the nupic library to run on 32bit Linux so it can run on raspberry PI.  That makes since because building a system with 100 raspberry PI modules is a cheap way to get lots of threads running in parallel.
If it does not affect our 64bit code performance I would like to also be able to compile as 32bit for this purpose if we can.

File: CommonCompilerConfig.cmake (hunk 6 above)
Comment (dkeeney) [reply CommonCompilerConfig.cmake]: > wouldn't the `-marm` be auto recognized by cmake?

good point.

File: src/CMakeLists.txt (hunk 7 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: But do we really need it split?
If there was a reason to generate separate static libraries perhaps but I don't see any need.

File: src/CMakeLists.txt (hunk 8 above)
Comment (dkeeney) [reply src/CMakeLists.txt]: I don't see the need.  But if you do want to split NetworkAPI into its own library then you need to move some files around.
- ntypes/MemParser.cpp/hpp  and MemStream.hpp are not part of NetworkAPI
- utils/ Watcher.cpp/hpp is part of NetworkAPI and not algorithms
- There are some files that belong in both libraries.

File: CommonCompilerConfig.cmake (hunk 5 above)
Comment (breznak) [reply CommonCompilerConfig.cmake]: I'll keep nta_arch where it makes sence

**Comment by ctrl-z-9000-times** [comment]
@breznak,
For this PR to move forward I would recommend:
*  Backing out the changes to Cells4 and doing that in a different PR, since they don't appear to pass CI, and they are off the topic of PR.  Try `git checkout master Cells4.cpp`
* Don't split up the build for NetworkAPI & Algorithms, just build it all at once.  IMO splitting up the build makes the CMakeFiles more complicated with very little benefit.  The NetworkAPI does no actual work (just setup, orchestration, & administration) so it should not matter what compile flags it uses.  

Anywhere near 20% improvement in SP without changing any code is great!

**Comment by breznak** [comment]
I'd like to merge this after #242 so that the changes can be validated on MNIST as well as on Hotgym.

**Comment by breznak** [comment]
I'll later merge master and see if there's anything worth left in the PR, or if we can close it.

**Comment by breznak** [comment]
Closing, most of this (compiler flags) has been merged by other PR, and I'm lazy resolving and finding what else is worth here, probably not much.
